Lib/concurrent/futures/process.py

```diff
@@ -400,11 +411,15 @@ def process_result_item(self, result_item):
         if isinstance(result_item, int):
             # Clean shutdown of a worker using its PID
             # (avoids marking the executor broken)
-            assert self.is_shutting_down()
             p = self.processes.pop(result_item)
             p.join()
-            if not self.processes:
-                self.join_executor_internals()
+            if self.is_shutting_down() and not self.pending_work_items:
+                if not self.processes:
+                    self.join_executor_internals()
+                    return
+            else:
+                if executor := self.executor_reference():
+                    executor._add_process_to_pool()
                 return
```

Comment:
The `return` statements are a bit pointless here since no other code would be executed otherwise.

Reply:
Agreed. I put them because there was a useless `return` before. I am happy to remove them.